gdxpds/special.py

```diff
@@ -0,0 +1,223 @@
+
+    Returns
+    -------
+    pandas.DataFrame
+        copy of df for which all numpy special values have been converted to
+        their GDX equivalents
+    """
+
+    # converts a single value; NANs are assumed already handled
+    def convert_approx_eps(value):
+        # eps values are not always caught by ==, use is_np_eps which applies
+        # a tolerance
+        if is_np_eps(value):
+            return SPECIAL_VALUES[4]
+        return value
+
+    # get a clean copy of df
+    try:
+        tmp = copy.deepcopy(df)
+    except:
+        logger.warning("Unable to deepcopy:\n{}".format(df))
+        tmp = copy.copy(df)
+
+    # fillna and apply map to value columns, then merge with dimensional columns
+    try:
+        values = tmp.iloc[:, num_dims:].replace(NP_TO_GDX_SVS, value=None).applymap(convert_approx_eps)
+        tmp = (tmp.iloc[:, :num_dims]).merge(values, left_index=True, right_index=True)
+    except:
+        logger.error("Unable to convert numpy special values to GDX special values." + \
+                     "num_dims: {}, dataframe:\n{}".format(num_dims, df))
+        raise
+    return tmp
+
+
+def pd_isnan(val):
+    """
+    Utility function for identifying None or NaN (which are indistinguishable in pandas).
+
+    Parameters
+    ----------
+    val : numeric
+        value to test
+
+    Returns
+    -------
+    bool
+        True if val is a GDX encoded special value that maps to None or numpy.nan;
+        False otherwise
+    """
+    return val is None or val != val
+
+
+def pd_val_equal(val1, val2):
+    """
+    Utility function used to test special value conversions.
+
+    Parameters
+    ----------
+    val1 : float or None
+        first value to compare
+    val2 : float or None
+        second value to compare
+
+    Returns
+    -------
+    bool
+        True if val1 and val2 are equal in the sense of == or they are
+        both NaN/None. The values that map to None and np.nan are assumed
+        to be equal because pandas cannot be relied upon to make the
+        distinction.
+    """
+    return pd_isnan(val1) and pd_isnan(val2) or val1 == val2
+
+
+def special_getter():
+    H = gdxcc.new_gdxHandle_tp()
+    rc = gdxcc.gdxCreateD(H, None, gdxcc.GMS_SSSIZE)
+    if not rc:
+        raise Exception(rc[1])
+    # get special values
+    special_values = gdxcc.doubleArray(gdxcc.GMS_SVIDX_MAX)
+    gdxcc.gdxGetSpecialValues(H, special_values)
+
+    special_list = []
+    gdx_to_np_svs = {}
+    np_to_gdx_svs = {}
+    for i in range(gdxcc.GMS_SVIDX_MAX):
+        if i >= len(NUMPY_SPECIAL_VALUES):
+            break
+        special_list.append(special_values[i])
+        gdx_val = special_values[i]
+        gdx_to_np_svs[gdx_val] = NUMPY_SPECIAL_VALUES[i]
+        np_to_gdx_svs[NUMPY_SPECIAL_VALUES[i]] = gdx_val
+
+    gdxcc.gdxFree(H)
+    return special_list, gdx_to_np_svs, np_to_gdx_svs
+
+
+SPECIAL_VALUES, GDX_TO_NP_SVS, NP_TO_GDX_SVS = special_getter()
```

Comment:
This is going to fail if someone needs to specify gams_dir. @jebob -- This is the biggest issue I see so far with the refactor.

Reply:
I think this is now fine, importing this module now initialises `SPECIAL_VALUES, GDX_TO_NP_SVS, NP_TO_GDX_SVS` as empty, they are populated when load_gdxcc finishes.

Is this satisfactory?